File: unity-solana-template/src/controllers/encrypt/PermutationObfuscate.ts
import {base64ToByteArray, byteArrayToBase64} from "../../utils/String.ts";
import IObfuscate from "./IObfuscate.ts";

export default class PermutationObfuscate implements IObfuscate {
    constructor(
        private readonly _permutationOrder: () => number[]
    ) {
    }

    obfuscate(base64: string): string {
        return this.obfuscate2(base64ToByteArray(base64));
    }

    obfuscate2(bytes: Uint8Array): string {
        const obfuscatedBytes = this.permuteKey(bytes, this._permutationOrder());
        return byteArrayToBase64(obfuscatedBytes);
    }

    deobfuscate(base64: string): string {
        return byteArrayToBase64(this.deobfuscate2(base64));
    }

    deobfuscate2(base64: string): Uint8Array {
        const bytes = base64ToByteArray(base64);
        return this.reversePermutation(bytes, this._permutationOrder());
    }

    private permuteKey(key: Uint8Array, permutationOrder: number[]): Uint8Array {
        if (key.length < permutationOrder.length) {
            throw new Error("Invalid key length.");
        }

        const permutedKey = new Uint8Array(key.length);
        for (let i = 0; i < permutationOrder.length; i++) {
            permutedKey[i] = key[permutationOrder[i]];
        }
        for (let i = permutationOrder.length; i < key.length; i++) {
            permutedKey[i] = key[i];
        }
        return permutedKey;
    }

    private reversePermutation(key: Uint8Array, permutationOrder: number[]): Uint8Array {
        if (key.length < permutationOrder.length) {
            throw new Error("Invalid key length.");
        }

        const originalKey = new Uint8Array(key.length);
        for (let i = 0; i < permutationOrder.length; i++) {
            originalKey[permutationOrder[i]] = key[i];
        }
        for (let i = permutationOrder.length; i < key.length; i++) {
            originalKey[i] = key[i];
        }
        return originalKey;
    }
}

File: Assets/Share/Scripts/Communicate/Encrypt/PermutationObfuscate.cs
using System;

namespace Communicate.Encrypt {
    /**
    * Hoán vị bytes theo một permutation order cho trước.
    * Yêu cầu input string phải >= permutation order.
    * Nếu input string dài hơn permutation order, thì chỉ hoán vị những bytes đầu tiên.
    */
    public class PermutationObfuscate : IObfuscate {
        private readonly byte[] _permutationOrder;

        public PermutationObfuscate(byte[] permutationOrder) {
            _permutationOrder = permutationOrder;
        }

        public string Obfuscate(string base64) {
            return Obfuscate2(ByteUtils.Base64ToByteArray(base64));
        }

        public string Obfuscate2(byte[] bytes) {
            var obfuscatedBytes = PermuteKey(bytes, _permutationOrder);
            return ByteUtils.ByteArrayToBase64(obfuscatedBytes);
        }

        public string DeObfuscate(string base64) {
            return ByteUtils.ByteArrayToBase64(DeObfuscate2(base64));
        }

        public byte[] DeObfuscate2(string base64) {
            var bytes = ByteUtils.Base64ToByteArray(base64);
            return ReversePermutation(bytes, _permutationOrder);
        }

        public byte[] DeObfuscateBytes(byte[] data) {
            return ReversePermutation(data, _permutationOrder);
        }

        private byte[] PermuteKey(byte[] key, byte[] permutationOrder) {
            if (key.Length < permutationOrder.Length) {
                throw new ArgumentException("Invalid key length.");
            }

            var permutedKey = new byte[key.Length];
            for (var i = 0; i < permutationOrder.Length; i++) {
                permutedKey[i] = key[permutationOrder[i]];
            }

            for (var i = permutationOrder.Length; i < key.Length; i++) {
                permutedKey[i] = key[i];
            }

            return permutedKey;
        }

        private byte[] ReversePermutation(byte[] key, byte[] permutationOrder) {
            if (key.Length < permutationOrder.Length) {
                throw new ArgumentException("Invalid key length.");
            }

            var originalKey = new byte[key.Length];
            for (var i = 0; i < permutationOrder.Length; i++) {
                originalKey[permutationOrder[i]] = key[i];
            }

            for (var i = permutationOrder.Length; i < key.Length; i++) {
                originalKey[i] = key[i];
            }

            return originalKey;
        }
    }
}